### DISCUSSION TITLE
Delegate to specialized function on CPU

### DISCUSSION THREAD

**Discussion by AdrienCorenflos** [jax-ml/jax#24584] [Q&A] [answered]
Hi,

I have a function `func(a)` for which I have a specialized cython implementation when on CPU. What is the recommended recipe to branch evaluation between CPU and GPU for this?
I can see that the base code is using "lowering" to either CPU/GPU/TPU but I am not sure how to do this for a simple function.

For the sake of the example, we can assume that I have two implementations `func_cpu(a)` and `func_gpu(a)` and I'm looking to branch from one to the other automatically based on where `a` is.

Note: I do not care about gradients here.

Thanks!

**Comment by jakevdp** [comment]
Unfortunately we don't have a public API for this, though we've discussed in the past the possibility of providing a `custom_lowering` mechanism similar to `custom_jvp` (cc/ @dfm).

The mechanism we do have for this is to define a primitive, and register device-specific lowering rules for that primitive. This is a bit involved and requires some non-public APIs, but a simple example in JAX's codebase is the `spsolve_p` primitive: https://github.com/jax-ml/jax/blob/7c4cc9552ce625ad699488e3930ae5f2850ee84c/jax/experimental/sparse/linalg.py#L592-L598

Note that it has different lowerings on CPU and on GPU. `mlir.lower_fun` can be useful for converting Python implementations into mlir lowerings that you can register.

You can find more discussion of implementing custom primitives at https://jax.readthedocs.io/en/latest/jax-primitives.html

**Reply by dfm** [reply, marked as answer]
It sounds to me like [`jax.lax.platform_dependent`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.platform_dependent.html) might actually do what you need here. For example, the example from the original post might look like the following:

```python
def func(a):
  return jax.lax.platform_dependent(a, cpu=func_cpu, cuda=func_gpu)
```

**Reply by jakevdp** [reply]
Wow, I'd somehow never encountered that. +1

**Reply by AdrienCorenflos** [reply]
Neat! Thanks @dfm you saved me a lot of pain :D

**Reply by AdrienCorenflos** [reply]
FYI, this question was related to branching [`jnp.searchsorted`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.searchsorted.html#jax-numpy-searchsorted) automatically as per this note:

![image](https://github.com/user-attachments/assets/a9811cc4-97a3-4d93-8290-df86eb38cd04)

(not exactly this because I have a very specific case where both my input array `a` and the values I'm searching therein `v` are sorted).
